content/en/docs/tasks/tls/manual-rotation-of-ca-certificates.md

```diff
@@ -0,0 +1,137 @@
+
+    *   Make sure coredns, kube-proxy and other pods using in-cluster configs are working as expected.
+
+1. Append new CA to the file against `--client-ca-file` flag in the `kube-apiserver` and `kube-scheduler` configuration.
+
+    {{< note >}}
+
+    If you used `kubeadm` to deploy your cluster, the control plane components run as {{< glossary_tooltip text="static Pods" term_id="static-pod" >}}, managed by the {{< glossary_tooltip text="kubelet" term_id="kubelet" >}}.
+
+    {{< /note >}}
+
+1. Update certificates for user accounts by replacing the content of `client-certificate-data` and `client-key-data` respectively and the `certificate-authority-data` section in all `admin.conf`, `controller-manager.conf`, `scheduler.conf` files with Base64-encoded new certificate data:
+
+    For information about creating certificates for individual user accounts, see [Configure certificates for user accounts](/docs/setup/best-practices/certificates/#configure-certificates-for-user-accounts).
+
+1. Restart API servers with the certificates (`apiserver.crt`, `apiserver-kubelet-client.crt`, and `front-proxy-client.crt`) signed by new CA. You can use the existing private keys or new private keys. If you changed the private keys then update these in the Kubernetes certificates directory as well.
+
+    Since the pod trusts both old + new CAs, there will be a momentarily disconnection after which the pod's kube client will reconnect to the new API server that uses the certificate signed by the new CA.
+
+    *    Restart Scheduler to use the new CAs.
+
+    *    Make sure control plane components logs no TLS errors.
+
+    {{< note >}}
+
+    To generate certificates and private keys for your cluster using the `openssl` command line tool, see [Certificates (`openssl`)](/docs/concepts/cluster-administration/certificates/#openssl).    
+    You can also use [`cfssl`](/docs/concepts/cluster-administration/certificates/#cfssl).
+
+    {{< /note >}}
+
+1. If your cluster is using bootstrap tokens to join nodes, update the ConfigMap `cluster-info` in the `kube-public` namespace with new CA.
+
+    ```shell
+    base64_encoded_ca="$(base64 /etc/kubernetes/pki/ca.crt)"
+
+    kubectl get cm/cluster-info --namespace kube-public -o yaml | \
+        /bin/sed "s/\(certificate-authority-data:\).*/\1 ${base64_encoded_ca}" | \
+        kubectl apply -f -
+    ```
+
+1. Update the file against `clientCAFile` in kubelet configuration and `certificate-authority-data` in kubelet.conf to use the new CA on all nodes.
+
+    If your kubelet is not using client certificate rotation update `client-certificate-data` and `client-key-data` in kubelet.conf on all nodes along with the kubelet client certificate file usually found in `/var/lib/kubelet/pki`.
+
+1. Follow these steps on all nodes in a rolling fashion.
+
+    1. Restart kubelet to start using the new CA
```

Comment:
This could be more detailed. What constitutes "using the new CA"?

Reply:
Any suggestions with better words to include "At this point the kubelet will start using new CA to establish trusted communication with API server".

Reply:
There are 3 configurable kubelet certificate settings, and we should be clear about each of them:

* kubelet authentication configuration (the `clientCAFile` in the config file). The clientCA file needs to be updated before the controller manager starts signing new certs on the CSR API.
* Kubelet client cert (optional). This must be done only after the KCM starts using the new CA private key
* Kubelet server cert. This must be done only after the API server has updated `--kubelet-client-certificate` with a certificate signed by the new CA and `--kubelet-certificate-authority` includes both the old and new CAs

Reply:
@micahhausler  This is a bit confusing here, Both kubelet and API server would need to be restarted more than twice in the process. 
1. I agree the kubelet needs to be updated first with `clientCAFile` containing both old and new CA but to make that effective kubelet must be restarted.
2. The API server will be restarted once after the KCM, Aggregated APIs and kubelet (ClientCAFile + Kubelet Client certificate)are updated.
3. Kubelet's Server certificate are updated and the kubelet is restarted to make it effective.
4. Then `--kubelet-client-certificate` needs to be updated in APIserver and it needs restart again.
5. All components must remove Old CA and restart again.

Am I missing an important point here, it is possible for me to confuse with a lot of discussion happened here, please advise.